corehq/motech/repeaters/models.py

```diff
@@ -800,7 +817,8 @@ def already_processed():
             # the fact Couch uses optimistic locking to prevent one process
             # from overwriting the work of another with stale data.
             return
-        process_repeat_record.delay(self)
+        if self.redis_lock.acquire():
```

Comment:
Should this lock be acquired before `self.save()` is called? All of the existing locks are based on revisions of the record that have not been updated with the new `next_check` value, so `self.save()` will unlock the record because of

https://github.com/dimagi/commcare-hq/blob/776cdc50385a9669cae1c3f8673dbcb2abbc53e9/corehq/motech/repeaters/models.py#L785-L786

I had initially thought we may need to acquire it before _and after_ the save call, but I don't think it's necessary since the `next_check` date will be set in the future by `self.save()`.

Reply:
A possibly better option would be to move `if self.redis_lock.acquire():` back into `check_repeaters` before

https://github.com/dimagi/commcare-hq/blob/776cdc50385a9669cae1c3f8673dbcb2abbc53e9/corehq/motech/repeaters/tasks.py#L54

I think that's the only place the lock is needed.

Reply:
I think your initial thought is correct; It will need to block before `save()` otherwise it will never block, and it will need to be acquired after `save()` otherwise it won't block the next attempt.

I think it should be inside `attempt_forward_now()` because it also needs to be acquired when the repeat record is registered.

Urgh this is ugly. I'm glad it's temporary.

Reply:
Context from offline discussion: @millerdev is correct; we only need to check the locks of repeat records whose `next_check` has not been set in the future, which registering and saving will do.

Reply:
https://github.com/dimagi/commcare-hq/pull/23580/commits/09d4c092a6c73cdb8034cb2b2002ae0e2941a50b

Reply:
Nice! :tada:

Reply:
Thank you!